fix(transformer/arrow-functions): do not transform super that inside nested non-async method

Dunqing opened the pull request:
This `super.value` belongs to the nested class, we shouldn't transform it.
```js
class Outer {
    async method() {
      class Inner extends Outer {
        normal() {
          // `super.value` should not be transformed, because it is not in an async method
          super.value
        }
      }
    }
 }
```

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8335?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#8335** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8335?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8335?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8331** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8331?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8335).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing commented:
```shell
➜ just test-transform
cargo run -p oxc_transform_conformance -- --exec 
    Finished `dev` profile [unoptimized] target(s) in 0.46s
     Running `target/debug/oxc_transform_conformance --exec`
thread 'main' panicked at crates/oxc_data_structures/src/stack/non_empty.rs:276:13:
Cannot pop all entries
```

caused panic in https://github.com/oxc-project/oxc/blob/bd591d896a9ed075dc2e7f0231c513547a517fe7/crates/oxc_data_structures/src/stack/sparse.rs#L102. `has_values` is a `NonEmptyStack`, so it must contain a value. It's correct in `NonEmptyStack`, but not in `SparrseStack`. See @overlookmotel

overlookmotel commented on crates/oxc_transformer/src/common/arrow_function_converter.rs:
```diff
@@ -208,9 +208,33 @@ impl<'a> Traverse<'a> for ArrowFunctionConverter<'a> {
+        if self.is_async_only()
+            && (func.r#async ||
+            // is used to confirm if it is a nested class method.
+            self.super_methods_stack.len() != 0)
```
`self.super_methods_stack.len() != 0` is always true. `SparseStack<T>` is essentially the same as `NonEmptyStack<Option<T>>` - it's never empty.

Possibly this is the cause of the panic. But please see also my comment on #8331.

By the way, I'll admit the difference between stack types is confusing! I've tried to improve the documentation in #8356.

Dunqing replied:
Thank you for spending time to write a book！Yes here `pop` times more than `push`, because `pop` missing a async check. 

I'm doing this because I'm following the same logic as before, and it's safe to pop here because it only pops when is_class_method_like is true. Anyway, follow the principle is always right. I know how to fix it, just move the pop logic to exit_function, and pass `super_methods` to `insert_variable_statement_at_the_top_of_statements`. 

Another thing I want to avoid is passing too many arguments to `insert_variable_statement_at_the_top_of_statements`, we have to pass all stack values in call site, because only program we need to use `last_mut`, otherwise we use `pop`.

overlookmotel reviewed (approved):
Glad it's now working! I'll open an issue about some suggested changes.

overlookmotel commented:
### Merge activity

* **Jan 9, 6:40 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
fix(transformer/arrow-functions): do not transform super that inside nested non-async method (#8335)

This `super.value` belongs to the nested class, we shouldn't transform it.
```js
class Outer {
    async method() {
      class Inner extends Outer {
        normal() {
          // `super.value` should not be transformed, because it is not in an async method
          super.value
        }
      }
    }
 }
```
